kotlin/tests/src/test/kotlin/com/squareup/moshi/kotlin/reflect/KotlinJsonAdapterTest.kt

```diff
@@ -642,14 +642,27 @@ class KotlinJsonAdapterTest {
     }
   }
 
-  @Test fun objectDeclarationsNotSupported() {
+  @Test fun objectDeclarationsSupported() {
     val moshi = Moshi.Builder().add(KotlinJsonAdapterFactory()).build()
+    val adapter = moshi.adapter<ObjectDeclaration>()
+    assertThat(adapter.toJson(ObjectDeclaration)).isEqualTo("{}")
+    assertThat(adapter.fromJson("{}")).isSameAs(ObjectDeclaration)
+  }
+
+  @Test fun objectDeclarations_withJsonContent() {
+    val moshi = Moshi.Builder().add(KotlinJsonAdapterFactory()).build()
+    val adapter = moshi.adapter<ObjectDeclaration>()
+    assertThat(adapter.fromJson("""{"a":6}""")).isSameAs(ObjectDeclaration)
```

Comment:
The unfortunate part about this is that in this case reading and writing is now asymmetric. Since we can't hide the ignored content somewhere in the instance if you read this from an external source, then try to send it back, they'll get something different then what you started with.  

I'm not sure how that could be mitigated unless there was a requirement that the `ObjectDeclaration` conformed to some sort of interface that allowed storing that extra data.  So the question is, is that important?  If so, then I think we should fail in this case, only supporting empty sentinel types.  If not, we can ignore it.

Reply:
Unless, of course, that's what `failOnUnknown()` is for.  I'm not familiar with that.

Reply:
When I talked about this idea with @swankjesse some time back, the idea was that `failOnUnknown()` would be usable to enforce this. That would break the example in this case, but that's a separate [known issue](https://github.com/square/moshi/pull/925). The only reason not to just immediately skip processing the keys entirely (which are always skipped here) is so that `failOnUnknown()` can catch them. If we can resolve #925 (Jesse has an idea for this), then we could even just always enable that in this adapter's scope.

As for asymmetric - I think it's only asymmetric at the JSON level and not the Moshi/adapter level. There's a precedent for this already for unknown JSON keys (i.e. enable `failOnUnknown()` if you care) and transient properties for data that never make it to the written JSON.